[oneDPL] Add memory parallel range algorithms (#631)
uxlfoundation/oneAPI-spec: source/elements/oneDPL/source/parallel_api/parallel_range_api.rst

.. code:: cpp

   // C++20 analogue of std::projected_value_t; exposition only
   template <typename I, typename Proj>
   using /*projected-value-type*/ = std::remove_cvref_t<std::invoke_result_t<Proj&, std::iter_value_t<I>&>>;

>   // C++20 analogue of nothrow-random-access-range proposed for C++26 in P3179R9; exposition only

Reviewer: ```suggestion
  // C++20 analogue of nothrow-random-access-range in the C++26 working draft; exposition only
  // Semantic requirements are listed further below
```

  template <typename R>

>   concept nothrow-random-access-range =

Reviewer 2: why did we decide to "copy" this concept instead of *`nothrow-sized-random-access-range`*?

    std::ranges::random_access_range<R> &&
    std::is_lvalue_reference_v<std::iter_reference_t<std::ranges::iterator_t<R>>> &&
    std::same_as<std::remove_cvref_t<std::iter_reference_t<std::ranges::iterator_t<R>>>,
                 std::iter_value_t<std::ranges::iterator_t<R>>>;

> Semantic Requirements
> ~~~~~~~~~~~~~~~~~~~~~
> 

Reviewer: I do not think an extra header is needed.
```suggestion
```

A type ``R`` models ``nothrow-random-access-range`` if no exceptions are thrown from:

> - increment, decrement, copy construction, move construction, copy assignment, move assignment,
>   comparisons or indirection through valid iterators of type ``std::ranges::iterator_t<R>``;
> - ``-`` operator, copy construction, move construction, copy assignment, move assignment,
>   or comparisons between valid values of type ``std::ranges::iterator_t<R>``;
> - ``-``, ``+``, ``-=``, ``+=``, ``[]`` operators on valid values of type
>   ``std::ranges::iterator_t<R>`` and ``std::iter_difference_t<std::ranges::iterator_t<R>>``;

Reviewer: ```suggestion
- any operation on an object of type ``std::ranges::iterator_t<R>`` required by the ``std::random_access_iterator`` concept;
```
While the standard says that some operations on no-throw iterators are allowed to throw, after close examination the only one I found not covered by the list is the default constructor. I think we can either ignore it or mention as an exception.

.. code:: cpp

   // C++20 analogue of std::projected_value_t; exposition only
   template <typename I, typename Proj>
   using /*projected-value-type*/ = std::remove_cvref_t<std::invoke_result_t<Proj&, std::iter_value_t<I>&>>;

  // C++20 analogue of nothrow-random-access-range proposed for C++26 in P3179R9; exposition only

>   template <typename R>
>   concept nothrow-random-access-range =
>     std::ranges::random_access_range<R> &&
>     std::is_lvalue_reference_v<std::iter_reference_t<I>> &&
>     std::same_as<std::remove_cvref_t<std::iter_reference_t<I>>, std::iter_value_t<I>> &&
>     std::sized_sentinel_for<std::ranges::sentinel_t<R>, std::ranges::iterator_t<R>>;

Reviewer 3: I think you missed extracting the iterator from the range when making this exposition only concept
```suggestion
  template <typename R>
  concept nothrow-random-access-range =
    std::ranges::random_access_range<R> &&
    std::is_lvalue_reference_v<std::iter_reference_t<std::ranges::iterator_t<R>>> &&
    std::same_as<std::remove_cvref_t<std::iter_reference_t<std::ranges::iterator_t<R>>>, std::iter_value_t<std::ranges::iterator_t<R>>> &&
    std::sized_sentinel_for<std::ranges::sentinel_t<R>, std::ranges::iterator_t<R>>;
```

   template<typename I, typename Proj>
   using /*projected-value-type*/ = std::remove_cvref_t<std::invoke_result_t<Proj&, std::iter_value_t<I>&>>;

  // Extension of C++20 exposition-only special memory concepts as defined in [special.mem.concepts]
  template<typename S, typename I>
  concept no-throw-sized-sentinel-for = // exposition only
    no-throw-sentinel-for<S, I> &&
    std::sized_sentinel_for<S, I>;

  template<typename I>
  concept no-throw-bidirectional-iterator = // exposition only
    no-throw-forward-iterator<I> &&
    std::bidirectional_iterator<I>;

  template<typename I>
  concept no-throw-random-access-iterator = // exposition only
    no-throw-bidirectional-iterator<I> &&
    std::random_access_iterator<I> &&
    no-throw-sized-sentinel-for<I, I>;

  template<typename R>
  concept no-throw-bidirectional-range = // exposition only
    no-throw-forward-range<R> &&
    no-throw-bidirectional-iterator<std::ranges::iterator_t<R>>;

  template<typename R>
  concept no-throw-random-access-range = // exposition only
    no-throw-bidirectional-range<R> &&

>     no-throw-random-access-iterator<std::ranges::iterator_t<R>>;

Author: Here is a "collapsed" representation `no-throw-random-access-range`, which I got by expanding each nested concept and removing already included concepts:

```c++
template<typename R>
concept no-throw-random-access-range = // exposition only
  ranges::range<R> &&
  std::random_access_iterator<std::ranges::iterator_t<R>> &&
  std::is_lvalue_reference_v<std::iter_reference_t<std::ranges::iterator_t<R>>> &&
  std::same_as<std::remove_cvref_t<std::iter_reference_t<std::ranges::iterator_t<R>>>,
               std::iter_value_t<std::ranges::iterator_t<R>>> &&
  std::sized_sentinel_for<std::ranges::sentinel_t<R>, std::ranges::iterator_t<R>>;
```

I suppose it is better to keep the original and more similar to the standard representation, especially because the nested concepts may also have syntactic meaning, where `no-throw-santinel-for` a good example with its pure syntactic nature (it's expanded only to `std::sentinel_for`).

   template<typename I, typename Proj>
   using /*projected-value-type*/ = std::remove_cvref_t<std::invoke_result_t<Proj&, std::iter_value_t<I>&>>;

  // Extension of C++20 exposition-only special memory concepts as defined in [special.mem.concepts]
  template<typename S, typename I>
  concept no-throw-sized-sentinel-for = // exposition only
    no-throw-sentinel-for<S, I> &&
    std::sized_sentinel_for<S, I>;

  template<typename I>
  concept no-throw-bidirectional-iterator = // exposition only
    no-throw-forward-iterator<I> &&

>     std::bidirectional_iterator<I>;

Reviewer: I think defining 7 exposition-only concepts (yes, technically two more are needed - for sentinels and forward iterators), of which only one is used in the algorithm signatures, is an overkill.

Perhaps I would only add `no-throw-sized-random-access-range` as the combination of sized and random access ranges with additional semantical requirements described by words.

Tagging @rarutyun for another opinion.

               std::ranges::sized_range<IR> && std::ranges::sized_range<OR> &&
               std::constructible_from<std::ranges::range_value_t<OR>, std::ranges::range_reference_t<IR>>
      std::ranges::uninitialized_copy_result<std::ranges::borrowed_iterator_t<IR>,
                                             std::ranges::borrowed_iterator_t<OR>>
        uninitialized_copy (ExecutionPolicy&& pol, IR&& in_range, OR&& out_range);

    // uninitialized_move
    template <typename ExecutionPolicy, std::ranges::random_access_range IR,
              /*nothrow-random-access-range*/ OR>
      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
               std::ranges::sized_range<IR> && std::ranges::sized_range<OR> &&
               std::constructible_from<std::ranges::range_value_t<OR>, std::ranges::range_rvalue_reference_t<IR>>
      std::ranges::uninitialized_move_result<std::ranges::borrowed_iterator_t<IR>,
                                             std::ranges::borrowed_iterator_t<OR>>
        uninitialized_move (ExecutionPolicy&& pol, IR&& in_range, OR&& out_range);

    // uninitialized_fill

>     template <typename ExecutionPolicy, /*nothrow-random-access-range*/ R,
>               typename T>

Reviewer: ```suggestion
    template <typename ExecutionPolicy, /*nothrow-random-access-range*/ R, typename T>
```

               std::constructible_from<std::ranges::range_value_t<OR>, std::ranges::range_reference_t<IR>>
      std::ranges::uninitialized_move_result<std::ranges::borrowed_iterator_t<IR>,
                                             std::ranges::borrowed_iterator_t<OR>>
        uninitialized_move (ExecutionPolicy&& pol, IR&& in_range, OR&& out_range);

    // uninitialized_fill
    template <typename ExecutionPolicy, /*no-throw-random-access-range*/ R,
              typename T>
      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
               std::ranges::sized_range<R> &&
               std::constructible_from<std::ranges::range_value_t<R>, const T&>
      std::ranges::borrowed_iterator_t<R>
        uninitialized_fill (ExecutionPolicy&& pol, R&& r, const T& value);

    // destroy
    template <typename ExecutionPolicy, /*no-throw-random-access-range*/ R>
      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
               std::ranges::sized_range<R> &&
               std::destructible<std::ranges::range_value_t<R>>
      std::ranges::borrowed_iterator_t<R>

>         destroy (ExecutionPolicy&& pol, R&& r);

Author: `destroy` is `noexcept` in https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2025/p3179r9.html, but I have added it here. @akukanov, could you confirm that is the correct or describe why it is not? 

Here are my considerations. Copy and move constructors of oneDPL executions policies are not necessary `noexcept`. Also, underlying parallel execution utilities can throw exceptions (e.g. SYCL parallel_for).

Reviewer: Tagging @rarutyun.
I think it is correct not to add `noexcept` in our case. This needs then to be described as another deviation from the standard at the beginning of the page.

                                       std::ranges::range_rvalue_reference_t<IR>>
      std::ranges::uninitialized_move_result<std::ranges::borrowed_iterator_t<IR>,
                                             std::ranges::borrowed_iterator_t<OR>>
        uninitialized_move (ExecutionPolicy&& pol, IR&& in_range, OR&& out_range);

    // uninitialized_fill
    template <typename ExecutionPolicy, /*nothrow-random-access-range*/ R,
              typename T = std::ranges::range_value_t<R>>
      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
               std::ranges::sized_range<R> &&
               std::constructible_from<std::ranges::range_value_t<R>, const T&>
      std::ranges::borrowed_iterator_t<R>
        uninitialized_fill (ExecutionPolicy&& pol, R&& r, const T& value);

    // destroy
    template <typename ExecutionPolicy, /*nothrow-random-access-range*/ R>
      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
               std::ranges::sized_range<R> &&
               std::destructible<std::ranges::range_value_t<R>>
      std::ranges::borrowed_iterator_t<R>

>         destroy (ExecutionPolicy&& pol, R&& r);
>   }

Reviewer: ```suggestion
        destroy (ExecutionPolicy&& pol, R&& r);

  }
```
To match the formatting of other code blocks.
